SCons/Environment.py

```diff
@@ -193,6 +193,162 @@ def _delete_duplicates(l, keep_last):
     return result
 
 
+def _add_cppdefines(
+    env_dict: dict,
+    val,  # add annotation?
+    prepend: bool = False,
+    unique: bool = False,
+    delete_existing: bool = False,
+) -> None:
+    """Adds to CPPDEFINES, using the rules for C preprocessor macros.
+
+    Split out from regular construction variable handling because these
+    entries can express either a macro with a replacement list or one
+    without.  A macro with replacement list can be supplied three ways:
+    as a combined string ``name=value``; as a tuple contained in
+    a sequence type ``[("name", value)]``; or as a dictionary entry
+    ``{"name": value}``.  Appending/prepending can be unconditional
+    (duplicates allowed) or uniquing (no dupes).
+
+    Note if a replacement list is supplied, "unique" requires a full
+    match - both the name and the replacement must be equal.
+
+    Args:
+        env_dict: the dictionary containing the ``CPPDEFINES`` to be modified.
+        val: the value to add, can be string, sequence or dict
+        prepend: whether to put *val* in front or back.
+        unique: whether to add *val* if it already exists.
+        delete_existing: if *unique* is true, add *val* after removing previous.
+    """
+
+    def _add_define(item, defines: deque, prepend: bool = False) -> None:
```

Comment:
not a fan of embedded functions..

Reply:
What can I say?  They have their uses. Obviously, this one can be unrolled. I think it makes it a bit more readable.